File: solution-python/029_divide_two_integers.py
"""
	029. Divide Two Integers
	@name: li jin
	@date: Mar 17, 2016
	@link: https://leetcode.com/problems/divide-two-integers/
	@time: 64 ms
"""

class Solution(object):

    def divide(self, dividend, divisor):
        MAX_INT = 0x7FFFFFFF
        if divisor == 0:
            return MAX_INT
        if dividend == 0:
            return 0
        negative = (dividend > 0) ^ (divisor > 0)
        dividend, divisor = abs(dividend), abs(divisor)
        ans = 0
        while dividend >= divisor:
            x = divisor
            i = 1
            while dividend >= x + x:
                x += x
                i += i
            dividend -= x
            ans += i
        ans = MAX_INT if not negative and ans > MAX_INT else ans
        return -ans if negative else ans

File: solution-cpp/029_divide_two_integers.cpp
/*
	029. Divide Two Integers
	@name: li jin
	@date: Mar 17, 2016
	@link: https://leetcode.com/problems/divide-two-integers/
	@time: 4 ms
*/

class Solution {
public:
    int divide(int dividend, int divisor) {
        if(dividend==0)  return 0;
        if(divisor==0)  return INT_MAX;

        double t1=log(fabs(dividend));
        double t2=log(fabs(divisor));
        long long result = double(exp(t1 - t2));
        if((dividend < 0) ^ (divisor < 0))  result = -result;
        if(result > INT_MAX)  result = INT_MAX;
        if(result < INT_MIN)  result = INT_MAX;
        return result;
    }
};
